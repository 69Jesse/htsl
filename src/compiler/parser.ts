import type { Lexer } from "./lexer.js";
import type { Delimiter, Token } from "./token.js";
import { span } from "../span.js";
import { error } from "../diagnostic.js";

export class Parser {

    lexer: Lexer;

    token: Token;
    prev: Token;

    constructor(lexer: Lexer) {
        this.lexer = lexer;
        this.token = { kind: "eof", span: span(0, 0) };
        this.prev = this.token;
    }

    parseStatName(): string {
        if (this.token.kind != "ident" && this.token.kind != "str") {
            throw error("expected stat name", this.token.span);
        }
        const value = this.token.value;
        this.next();
        return value;
    }

    parseI64(): BigInt {
        if (this.token.kind != "i64") {
            throw error("expected number", this.token.span);
        }
        const value = BigInt(this.token.value); // this should never fail
        this.next();

        if (value < BigInt("-9223372036854775808") || value > BigInt("9223372036854775807")) {
            throw error("number exceeds 64-bit integer limit", this.token.span);
        }
        return value;
    }

    parseF64(): number {
        if (this.token.kind != "i64" && this.token.kind != "f64") {
            throw error("expected number", this.token.span);
        }

        return parseFloat(this.token.value);
    }

    parseDelimitedCommaSeq<T>(delim: Delimiter, parser: () => T) {
        if (this.token.kind != "open_delim" || this.token.delim != delim) {
            throw error("expected (", this.token.span);
        }
        this.next();

        while (this.token.kind != "eof") {
            
        }
    }

    match(kind: Token["kind"]) {
        if (!this.eat(kind)) {
            throw error(`expected ${kind}`, this.token.span);
        }
    }

    eat(kind: Token["kind"]): boolean {
        const matches = this.token.kind == kind;
        if (matches) this.next();
        return matches;
    }

    next() {
        this.prev = this.token;
        this.token = this.lexer.advanceToken();
    }

}