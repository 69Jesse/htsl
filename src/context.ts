import type { Diagnostic } from "./parse/index.js";

export class CompileCtx {

    diagnostics: Array<Diagnostic>

    constructor() {
        this.diagnostics = [];
    }

    emit(diagnostic: Diagnostic) {
        this.diagnostics.push(diagnostic);
    }

}
