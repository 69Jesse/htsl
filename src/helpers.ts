// biome-ignore lint/suspicious/noExplicitAny: function should run on anything
export function partialEq(src: any, target: any): boolean {
	return Object.keys(target).every((key) => {
		return target[key] === src[key];
	});
}

export function isAction(name: string) {
	return [
		"applyLayout", "applyPotion", "balanceTeam", "cancelEvent", "changeHealth", "hungerLevel",
		"maxHealth", "changePlayerGroup", "clearEffects", "closeMenu", "actionBar", "displayMenu",
		"title", "enchant", "exit", "failParkour", "fullHeal", "xpLevel", "giveItem", "houseSpawn",
		"kill", "parkCheck", "pause", "sound", "removeItem", "resetInventory", "chat", "lobby",
		"compassTarget", "gamemode", "setTeam", "tp", "consumeItem", "stat",
	].includes(name);
}

export const ACTION_NAMES: {
	[key: string]: string
} = {
	"CHANGE_STAT": "stat",
	"TITLE": "title",
	"MESSAGE": "chat",
	"CHANGE_HEALTH": "changeHealth",
};