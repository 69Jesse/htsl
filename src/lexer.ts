import { token, type Token } from "./token.js";

class Lexer {

    src: string;
    pos: number;

    constructor(src: string, pos: number) {
        this.src = src;
        this.pos = pos;
    }

    advanceToken(): Token {
        // eat whitespace
        while (this.hasNext() && /^\s+$/.test(this.peek())) { this.next(); }
        if (!this.hasNext()) return token("eof");

        const lo = this.pos;
        const c = this.next();

        if (c == "/" && this.peek() == "/") {
            // eat line comment
            do {
                this.next();
            } while (this.hasNext() && this.peek() != "\n");

            return this.advanceToken();
        }

        if (c == "/" && this.peek() == "*") {
            this.next();

            // eat block comment
            let depth = 1;
            while (this.hasNext()) {
                const c = this.next();
                if (c == "/" && this.peek() == "*") {
                    this.next();
                    depth++;
                }
                else if (c == "*" && this.peek() == "/") {
                    this.next();
                    depth--;
                    if (depth == 0) break;
                }
            }

            return this.advanceToken();
        }

        
    }

    hasNext(): boolean {
        return this.pos < this.src.length;
    }
    
    next(): string {
        return this.src.charAt(this.pos++);
    }

    peek(skip?: number): string {
        return this.src.charAt(this.pos + (skip ?? 0));
    }

}