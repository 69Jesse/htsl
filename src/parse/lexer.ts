import { span } from "../types/span.js";
import { token, type Token } from "./token.js";
import { Diagnostic } from "../types/diagnostic.js";

export class Lexer {
	src: string;
	pos: number;

	constructor(src: string) {
		this.src = src;
		this.pos = 0;
	}

	advanceToken(): Token {
		// eat whitespace
		while (this.hasNext() && /^\s+$/.test(this.peek())) {
			this.next();
		}
		if (!this.hasNext()) return token("eof", span(this.pos, this.pos));

		const lo = this.pos;
		const singleSpan = span(lo, lo);
		const c = this.next();

		if (c === "/" && this.peek() === "/") {
			// eat line comment
			do {
				this.next();
			} while (this.hasNext() && this.peek() !== "\n");

			return this.advanceToken();
		}

		if (c === "/" && this.peek() === "*") {
			this.next();

			// eat block comment
			let depth = 1;
			while (this.hasNext()) {
				const c = this.next();
				if (c === "/" && this.peek() === "*") {
					this.next();
					depth++;
				} else if (c === "*" && this.peek() === "/") {
					this.next();
					depth--;
					if (depth === 0) break;
				}
			}

			return this.advanceToken();
		}

		if (c === ",") return token("comma", singleSpan);

		// binary operators
		if (c === "+") {
			if (this.peek(0) === "=") {
				this.next();
				return token("bin_op_eq", span(lo, lo + 1), { op: "plus" });
			}
			return token("bin_op", singleSpan, { op: "plus" });
		}
		if (c === "-") {
			if (this.peek(0) === "=") {
				this.next();
				return token("bin_op_eq", span(lo, lo + 1), { op: "minus" });
			}
			return token("bin_op", singleSpan, { op: "minus" });
		}
		if (c === "*") {
			if (this.peek(0) === "*") {
				this.next();
				return token("bin_op", span(lo, lo + 1), { op: "star_star" });
			}
			if (this.peek(0) === "=") {
				this.next();
				return token("bin_op_eq", span(lo, lo + 1), { op: "star" });
			}
			return token("bin_op", singleSpan, { op: "star" });
		}
		if (c === "/") {
			if (this.peek(0) === "/") this.next();
			if (this.peek(0) === "=") {
				this.next();
				return token("bin_op_eq", span(lo, this.pos), { op: "slash" });
			}
			return token("bin_op", span(lo, this.pos), { op: "slash" });
		}

		// comparison operators
		if (c === "=") {
			if (this.peek(0) === "=") {
				this.next();
				return token("cmp_op_eq", span(lo, lo + 1), { op: "equal" });
			}
			return token("cmp_op", singleSpan, { op: "equal" });
		}
		if (c === "<") {
			if (this.peek(0) === "=") {
				this.next();
				return token("cmp_op_eq", span(lo, lo + 1), { op: "less_than" });
			}
			return token("cmp_op", singleSpan, { op: "less_than" });
		}
		if (c === ">") {
			if (this.peek(0) === "=") {
				this.next();
				return token("cmp_op_eq", span(lo, lo + 1), { op: "greater_than" });
			}
			return token("cmp_op", singleSpan, { op: "greater_than" });
		}

		// delimiters
		if (c === "(")
			return token("open_delim", singleSpan, { delim: "parenthesis" });
		if (c === ")")
			return token("close_delim", singleSpan, { delim: "parenthesis" });
		if (c === "{") return token("open_delim", singleSpan, { delim: "brace" });
		if (c === "}") return token("close_delim", singleSpan, { delim: "brace" });
		if (c === "[") return token("open_delim", singleSpan, { delim: "bracket" });
		if (c === "]") return token("close_delim", singleSpan, { delim: "bracket" });

		// literals
		if (c === '"') {
			let value = "";
			let escapeNext = false;
			while (this.hasNext()) {
				const c = this.next();
				if (!escapeNext && c === '"') break;
				if (!escapeNext && c === "\\") {
					escapeNext = true;
					continue;
				}
				escapeNext = false;
				value += c;
			}

			return token("str", span(lo, this.pos), { value });
		}

		if (/[0-9]/.test(c)) {
			let value = c;
			while (this.hasNext()) {
				if (!/[0-9]/.test(this.peek(0))) break;
				value += this.next();
			}
			if (this.peek(0) === ".") {
				value += ".";
				while (this.hasNext()) {
					if (!/[0-9]/.test(this.peek(0))) break;
					value += this.next();
				}
				return token("f64", span(lo, this.pos), { value });
			}
			return token("i64", span(lo, this.pos), { value });
		}

		if (/[a-zA-Z_]/.test(c)) {
			let value = c;
			while (this.hasNext()) {
				if (!/[a-zA-Z_]/.test(this.peek(0))) break;
				value += this.next();
			}
			return token("ident", span(lo, this.pos), { value });
		}

		if (c === "\n") return token("eol", singleSpan);

		throw Diagnostic.error(`unknown token "${c}"`, span(lo, lo)); // Eventually turn this into a diagnostic
	}

	hasNext(): boolean {
		return this.pos < this.src.length;
	}

	next(): string {
		return this.src.charAt(this.pos++);
	}

	peek(skip?: number): string {
		return this.src.charAt(this.pos + (skip ?? 0));
	}
}
