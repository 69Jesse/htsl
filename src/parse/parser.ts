import type { CompileCtx } from "../context.js";
import { isAction, partialEq } from "../helpers.js";
import type {
	Action,
	ActionChangeGlobalStat,
	ActionChangeHealth,
	ActionChangeStat,
	ActionChangeTeamStat, ActionConditional,
	ActionMessage,
	ActionTitle
} from "../action/action.js";
import type { Lexer } from "./lexer.js";
import {
	type Delimiter,
	type I64Kind, type PlaceholderKind,
	type StrKind,
	type Token,
	tokenToString,
} from "./token.js";
import { type Span, span } from "../types/span.js";
import { Diagnostic } from "../types/diagnostic.js";
import type { Amount, Mode } from "../action/action.js";

export class Parser {
	ctx: CompileCtx;
	lexer: Lexer;

	token: Token;
	prev: Token;

	constructor(ctx: CompileCtx, lexer: Lexer) {
		this.ctx = ctx;
		this.lexer = lexer;
		this.token = { kind: "eof", span: span(0, 0) };
		this.prev = this.token;
		this.next();
	}

	parseCompletely(): Array<Action> {
		const actions = [];
		while (true) {
			try {
				const action = this.parseAction();
				if (!action) break;
				actions.push(action);
			} catch (e) {
				if (e instanceof Diagnostic) this.ctx.emit(e);
				this.recover();
			}
		}
		return actions;
	}

	parseAction(): Action | undefined {
		if (this.eatIdent("stat")) {
			return this.parseActionChangeStat();
		}
		if (this.eatIdent("globalstat")) {
			return this.parseActionChangeGlobalStat();
		}
		if (this.eatIdent("teamstat")) {
			return this.parseActionChangeTeamStat();
		}
		if (this.eatIdent("changeHealth")) {
			return this.parseActionChangeHealth();
		}
		if (this.eatIdent("title")) {
			return this.parseActionTitle();
		}
		if (this.eatIdent("chat")) {
			return this.parseActionMessage();
		}
		if (this.eatIdent("if")) {
			return this.parseActionConditional();
		}

		if (this.check("ident")) {
			throw Diagnostic.error("Unknown action", this.token.span);
		}
		if (!this.check("eof")) {
			throw Diagnostic.error("Expected action", this.token.span);
		}
		return undefined;
	}

	parseActionChangeStat(): ActionChangeStat {
		let stat, mode, amount;
		this.parseRecovering(() => {
			stat = this.parseSpanned(() => this.parseStatName());
			mode = this.parseSpanned(() => this.parseStatMode());
			amount = this.parseSpanned(() => this.parseStatAmount());
		});

		return { type: "CHANGE_STAT", stat, mode, amount };
	}

	parseActionChangeGlobalStat(): ActionChangeGlobalStat {
		let stat, mode, amount;
		this.parseRecovering(() => {
			stat = this.parseSpanned(() => this.parseStatName());
			mode = this.parseSpanned(() => this.parseStatMode());
			amount = this.parseSpanned(() => this.parseStatAmount());
		});

		return { type: "CHANGE_GLOBAL_STAT", stat, mode, amount };
	}

	parseActionChangeTeamStat(): ActionChangeTeamStat {
		let stat, team, mode, amount;
		this.parseRecovering(() => {
			stat = this.parseSpanned(() => this.parseStatName());
			team = this.parseSpanned(() => this.parseStatName());
			mode = this.parseSpanned(() => this.parseStatMode());
			amount = this.parseSpanned(() => this.parseStatAmount());
		});

		return { type: "CHANGE_TEAM_STAT", stat, team, mode, amount };
	}

	parseActionChangeHealth(): ActionChangeHealth {
		let mode, health;
		this.parseRecovering(() => {
			mode = this.parseSpanned(() => this.parseStatMode());
			health = this.parseSpanned(() => this.parseStatAmount());
		});

		return { type: "CHANGE_HEALTH", mode, health };
	}

	parseActionTitle(): ActionTitle {
		let title, subtitle, fadein, stay, fadeout;
		this.parseRecovering(() => {
			title = this.parseSpanned(() => this.parseStr());
			subtitle = this.parseSpanned(() => this.parseStr());

			fadein = this.parseSpanned(() => this.parseI64());
			stay = this.parseSpanned(() => this.parseI64());
			fadeout = this.parseSpanned(() => this.parseI64());
		});

		return { type: "TITLE", title, subtitle, fadein, stay, fadeout };
	}

	parseActionMessage(): ActionMessage {
		let message;
		this.parseRecovering(() => {
			message = this.parseSpanned(() => this.parseStr());
		});

		return { type: "MESSAGE", message };
	}

	parseActionConditional(): ActionConditional {
		let mode, conditions, ifActions, elseActions;
		this.parseRecovering(() => {
			mode = this.parseSpanned(() => {
				this.eatIdent("and");
				return this.eatIdent("or");
			});

			conditions = this.parseSpanned(() => this.parseDelimitedCommaSeq("parenthesis", () => {
				return "";
			}));

			ifActions = this.parseSpanned(() => this.parseBlock());

			if (this.eatIdent("else")) {
				elseActions = this.parseSpanned(() => this.parseBlock());
			}
		});

		return { type: "CONDITIONAL", mode, conditions, ifActions, elseActions };
	}

	parseStatName(): string {
		if (this.token.kind !== "ident" && this.token.kind !== "str") {
			throw Diagnostic.error("Expected stat name", this.token.span);
		}
		const value = this.token.value;
		if (value.length > 16) {
			throw Diagnostic.error("Stat name exceeds 16-character limit", this.token.span);
		}
		if (value.length < 1) {
			// we check this because of: ""
			throw Diagnostic.error("Stat name cannot be empty", this.token.span);
		}
		if (value.includes(" ")) {
			throw Diagnostic.error("Stat name cannot contain spaces", this.token.span);
		}

		this.next();
		return value;
	}

	parseStatMode(): Mode {
		if (this.eatOption("increment") || this.eatOption("inc") || this.eat({ kind: "bin_op_eq", op: "plus" })) {
			return "increment";
		}
		if (this.eatOption("decrement") || this.eatOption("dec") || this.eat({ kind: "bin_op_eq", op: "minus" })) {
			return "decrement";
		}
		if (this.eatOption("multiply") || this.eatOption("mul") || this.eat({ kind: "bin_op_eq", op: "star" })) {
			return "multiply";
		}
		if (this.eatOption("divide") || this.eatOption("div") || this.eat({ kind: "bin_op_eq", op: "slash" })) {
			return "divide";
		}
		if (this.eatOption("set") || this.eat({ kind: "cmp_op", op: "equal" })) {
			return "set";
		}
		if (this.check("str") || this.check("ident")) {
			throw Diagnostic.error("Expected operation (increment, decrement, multiply, divide, set)", this.token.span);
		}
		throw Diagnostic.error("Expected operation", this.token.span);
	}

	parseStatAmount(): Amount {
		if (this.check("i64") || this.check({ kind: "bin_op", op: "minus" })) {
			return this.parseI64();
		}
		if (this.check("placeholder")) {
			this.next();
			return (this.prev as PlaceholderKind).value;
		}
		if (this.check("str")) {
			this.next();
			return (this.prev as StrKind).value;
		}
		if (this.eatIdent("stat")) {
			const name = this.parseStatName();
			return `%stat.player/${name}%`;
		}
		if (this.eatIdent("globalstat")) {
			const name = this.parseStatName();
			return `%stat.global/${name}%`;
		}
		if (this.eatIdent("teamstat")) {
			const name = this.parseStatName();

			if (!this.check("ident") && !this.check("str")) {
				throw Diagnostic.error("Expected team name", this.token.span);
			}
			const team = this.parseStatName();
			return `%stat.team/${name} ${team}%`;
		}
		throw Diagnostic.error("expected amount", this.token.span);
	}

	parseStr(): string {
		this.expect("str");
		return (this.prev as StrKind).value;
	}

	parseI64(): bigint {
		const negative = this.eat({ kind: "bin_op", op: "minus" });

		this.expect("i64");
		let value = BigInt((this.prev as I64Kind).value); // bigint constructor should never fail?
		if (negative) value *= -1n;

		if (
			value < BigInt("-9223372036854775808") ||
			value > BigInt("9223372036854775807")
		) {
			throw Diagnostic.error("Number exceeds 64-bit integer limit", this.prev.span);
		}
		return value;
	}

	parseF64(): number {
		if (this.token.kind !== "i64" && this.token.kind !== "f64") {
			throw Diagnostic.error("Expected number", this.token.span);
		}
		this.next();

		return Number.parseFloat(this.token.value);
	}

	parseRecovering(parser: () => void) {
		try {
			return parser();
		} catch (e) {
			if (e instanceof Diagnostic) {
				// cursed span creation? maybe I will do this one day
				/*
				if (this.token.kind === "ident" && isAction(this.token.value)) {
					const endPos = this.prev.span.hi;
					e.span = span(endPos, endPos);
				}
				 */
				this.ctx.emit(e);
				this.recover();

			}
			else throw e;
		}
	}

	parseBlock(): Array<Action> {
		const actions = [];
		this.expect({ kind: "open_delim", delim: "brace" });
		while (true) {
			if (this.check("eof")) throw Diagnostic.error("expected }");
			if (this.eat({ kind: "close_delim", delim: "brace" })) break;
			try {
				const action = this.parseAction();
				if (!action) break;
				actions.push(action);
			} catch (e) {
				if (e instanceof Diagnostic) this.ctx.emit(e);
				this.recover();
			}
		}
		return actions;
	}

	parseDelimitedCommaSeq<T>(delim: Delimiter, parser: () => T) {
		this.expect({ kind: "open_delim", delim });

		const seq: Array<T> = [];
		while (!this.check({ kind: "close_delim", delim })) {
			if (this.token.kind === "eof") break; // will catch the below `expect` and error gracefully

			seq.push(parser());

			if (!this.eat("comma")) break;
		}
		this.expect({ kind: "close_delim", delim });

		return seq;
	}

	parseSpanned<T>(parser: () => T): { value: T, span: Span } {
		const lo = this.token.span.lo;
		const value = parser();
		const hi = this.prev.span.hi;

		return { value, span: span(lo, hi) };
	}

	eatOption(value: string): boolean {
		if (this.eatIdent(value)) return true;
		return this.eat({ kind: "str", value });
	}

	eatIdent(value: string): boolean {
		return this.eat({ kind: "ident", value });
	}

	recover() {
		while (true) {
			if (
				this.check("eof")
				|| (this.token.kind === "ident" && isAction(this.token.value))
			) return;
			this.next();
		}
	}

	expect(tok: Token["kind"] | Partial<Token>) {
		if (!this.eat(tok)) {
			throw Diagnostic.error(`Expected ${tokenToString(tok)}`, this.token.span);
		}
	}

	eat(tok: Token["kind"] | Partial<Token>): boolean {
		const matches = this.check(tok);
		if (matches) this.next();
		return matches;
	}

	check(tok: Token["kind"] | Partial<Token>): boolean {
		return typeof tok === "string"
			? this.token.kind === tok // match just kind
			: partialEq(this.token, tok); // match partial token
	}

	next() {
		this.prev = this.token;
		this.token = this.lexer.advanceToken();
	}
}
