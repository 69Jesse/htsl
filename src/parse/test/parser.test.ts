import { describe, expect, it } from "vitest";
import { readCases } from "./helpers.js";
import { Lexer } from "../lexer.js";
import { Parser } from "../parser.js";
import { CompileCtx } from "../../context.js";

describe("Parser", () => {

    for (const test of readCases(__dirname + "/cases/parser/")) {
        it("parse " + test.name, () => {
            const ctx = new CompileCtx();
            const lexer = new Lexer(test.source);
            const parser = new Parser(ctx, lexer);
            const result = parser.parseCompletely();

            expect(result).toMatchSnapshot();
        });
    }

    for (const test of readCases(__dirname + "/cases/parser/failing/")) {
        it("error " + test.name, () => {
            const ctx = new CompileCtx();
            const lexer = new Lexer(test.source);
            const parser = new Parser(ctx, lexer);
            parser.parseCompletely();

            const diagnosticData = ctx.diagnostics.map(it => {
                return {
                    level: it.level,
                    message: it.message,
                    span: it.span,
                }
            });

            expect(diagnosticData.length).toBeGreaterThan(0);
            expect(diagnosticData).toMatchSnapshot();
        })
    }

});