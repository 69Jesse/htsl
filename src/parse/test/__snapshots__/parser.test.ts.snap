// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parser > error invalid stat actions 1`] = `
[
  {
    "level": "error",
    "message": "Expected stat name",
    "span": {
      "hi": 5,
      "lo": 5,
    },
  },
  {
    "level": "error",
    "message": "expected amount",
    "span": {
      "hi": 28,
      "lo": 19,
    },
  },
  {
    "level": "error",
    "message": "Expected operation",
    "span": {
      "hi": 38,
      "lo": 37,
    },
  },
  {
    "level": "error",
    "message": "Expected stat name",
    "span": {
      "hi": 44,
      "lo": 44,
    },
  },
]
`;

exports[`Parser > error invalid stat names 1`] = `
[
  {
    "level": "error",
    "message": "Stat name exceeds 16-character limit",
    "span": {
      "hi": 23,
      "lo": 5,
    },
  },
  {
    "level": "error",
    "message": "Stat name cannot be empty",
    "span": {
      "hi": 36,
      "lo": 34,
    },
  },
  {
    "level": "error",
    "message": "Stat name cannot contain spaces",
    "span": {
      "hi": 60,
      "lo": 48,
    },
  },
]
`;

exports[`Parser > error number above integer limit 1`] = `
[
  {
    "level": "error",
    "message": "Number exceeds 64-bit integer limit",
    "span": {
      "hi": 40,
      "lo": 9,
    },
  },
]
`;

exports[`Parser > parse global stats 1`] = `
[
  {
    "amount": {
      "span": {
        "hi": 16,
        "lo": 15,
      },
      "value": 5n,
    },
    "mode": {
      "span": {
        "hi": 13,
        "lo": 13,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 12,
        "lo": 11,
      },
      "value": "x",
    },
    "type": "CHANGE_GLOBAL_STAT",
  },
  {
    "amount": {
      "span": {
        "hi": 45,
        "lo": 33,
      },
      "value": "%stat.global/x%",
    },
    "mode": {
      "span": {
        "hi": 31,
        "lo": 31,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 30,
        "lo": 29,
      },
      "value": "y",
    },
    "type": "CHANGE_GLOBAL_STAT",
  },
  {
    "amount": {
      "span": {
        "hi": 78,
        "lo": 62,
      },
      "value": "%stat.global/y%",
    },
    "mode": {
      "span": {
        "hi": 60,
        "lo": 60,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 59,
        "lo": 58,
      },
      "value": "z",
    },
    "type": "CHANGE_GLOBAL_STAT",
  },
]
`;

exports[`Parser > parse player stats 1`] = `
[
  {
    "amount": {
      "span": {
        "hi": 10,
        "lo": 9,
      },
      "value": 5n,
    },
    "mode": {
      "span": {
        "hi": 7,
        "lo": 7,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 6,
        "lo": 5,
      },
      "value": "x",
    },
    "type": "CHANGE_STAT",
  },
  {
    "amount": {
      "span": {
        "hi": 27,
        "lo": 21,
      },
      "value": "%stat.player/x%",
    },
    "mode": {
      "span": {
        "hi": 19,
        "lo": 19,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 18,
        "lo": 17,
      },
      "value": "y",
    },
    "type": "CHANGE_STAT",
  },
  {
    "amount": {
      "span": {
        "hi": 54,
        "lo": 38,
      },
      "value": "%stat.player/y%",
    },
    "mode": {
      "span": {
        "hi": 36,
        "lo": 36,
      },
      "value": "set",
    },
    "stat": {
      "span": {
        "hi": 35,
        "lo": 34,
      },
      "value": "z",
    },
    "type": "CHANGE_STAT",
  },
]
`;
